File: android_java_hummus/src/main/java/com/github/klee0kai/hummus/adapterdelegates/delegate/viewbinding/ClassViewBindingAdapterDelegate.java
package com.github.klee0kai.hummus.adapterdelegates.delegate.viewbinding;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewbinding.ViewBinding;

import com.hannesdorfmann.adapterdelegates4.AdapterDelegate;

import java.util.List;

public abstract class ClassViewBindingAdapterDelegate<T, VB extends ViewBinding> extends AdapterDelegate<List<Object>> {

    private final Class<T> tClass;
    private final boolean forceTheSameClass;

    public ClassViewBindingAdapterDelegate(Class<T> tClass, boolean forceTheSameClass) {
        this.tClass = tClass;
        this.forceTheSameClass = forceTheSameClass;
    }

    public ClassViewBindingAdapterDelegate(Class<T> tClass) {
        this(tClass, false);
    }

    @Override
    protected boolean isForViewType(@NonNull List<Object> items, int position) {
        Object item = items.get(position);
        if (forceTheSameClass) {
            return tClass.equals(item.getClass());
        } else {
            return tClass.isInstance(item);
        }
    }

    @NonNull
    @Override
    protected ViewBindingHolder<VB> onCreateViewHolder(ViewGroup parent) {
        return new ViewBindingHolder<>(onCreateViewBinding(LayoutInflater.from(parent.getContext()), parent));
    }

    @Override
    protected void onBindViewHolder(@NonNull List<Object> objects, int i, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull List<Object> list) {
        onBindViewHolder((T) objects.get(i), i, (ViewBindingHolder<VB>) viewHolder);
    }


    protected abstract VB onCreateViewBinding(LayoutInflater inflater, ViewGroup parent);

    protected abstract void onBindViewHolder(T it, int pos, @NonNull ViewBindingHolder<VB> vh);

}

File: android_kotlin_hummus/src/main/kotlin/com/github/klee0kai/hummus/adapterdelegates/delegate/viewbinding/ClassViewBindingAdapterDelegate.kt
package com.github.klee0kai.hummus.adapterdelegates.delegate.viewbinding

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import androidx.viewbinding.ViewBinding
import com.hannesdorfmann.adapterdelegates4.AdapterDelegate


abstract class ClassViewBindingAdapterDelegate<T, VB : ViewBinding>(
    protected open val tClass: Class<T>,
    protected open val forceTheSameClass: Boolean = false,
    protected open val bindingCreator: (LayoutInflater, ViewGroup) -> VB
) : AdapterDelegate<List<Any>>() {

    override fun isForViewType(items: List<Any>, position: Int): Boolean {
        val item = items[position]
        return if (forceTheSameClass) {
            tClass == item.javaClass
        } else {
            tClass.isInstance(item)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup): ViewBindingHolder<VB> {
        return ViewBindingHolder(bindingCreator(LayoutInflater.from(parent.context), parent))
    }

    override fun onBindViewHolder(
        objects: List<Any>,
        pos: Int,
        viewHolder: ViewHolder,
        list: List<Any>
    ) {
        onBindViewHolder(objects[pos] as T, pos, viewHolder as ViewBindingHolder<VB>)
    }

    protected abstract fun onBindViewHolder(it: T, pos: Int, vh: ViewBindingHolder<VB>)


    companion object {

        fun <T, VB : ViewBinding> create(
            tClass: Class<T>,
            forceTheSameClass: Boolean = false,
            bindingCreator: (LayoutInflater, ViewGroup) -> VB,
            bind: ViewBindingHolder<VB>.(it: T, pos: Int) -> Unit,
        ): ClassViewBindingAdapterDelegate<T, VB> {
            return object :
                ClassViewBindingAdapterDelegate<T, VB>(tClass, forceTheSameClass, bindingCreator) {

                override fun onBindViewHolder(it: T, pos: Int, vh: ViewBindingHolder<VB>) {
                    bind.invoke(vh, it, pos)
                }

            }
        }

    }
}
